src/sagemaker_xgboost_container/algorithm_mode/dask_training.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License'). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the 'license' file accompanying this file. This file is
+# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import logging
+import os
+import socket
+from subprocess import Popen
+import sys
+import time
+
+import dask.dataframe as dask_dataframe
+from dask.dataframe import DataFrame
+from dask.dataframe import Series
+from dask.distributed import Client, wait
+from sagemaker_algorithm_toolkit import exceptions as exc
+import xgboost as xgb
+
+from sagemaker_xgboost_container.constants.xgb_constants import MODEL_NAME
+
+DASK_PATH = "/miniconda3/bin"
+SCHEDULER_PORT = "8786"
+
+WAIT_TIME_DAEMONS_STARTUP_SEC = 2
+WAIT_FOR_ALL_WORKERS_TIMEOUT_SEC = 20
+
+logger = logging.getLogger(__name__)
+
+
+def _start_daemons(scheduler_host_ip: str, current_host_ip: str):
+    cmd_start_scheduler = os.path.join(DASK_PATH, "dask-scheduler")
+    cmd_start_worker = os.path.join(DASK_PATH, "dask-cuda-worker")
+    schedule_conn_string = "tcp://{}:{}".format(scheduler_host_ip, SCHEDULER_PORT)
+
+    # We can set this based on number of vCPUs in the future.
+    # So far haven't seen improvement with > 4, during testing.
+    num_threads_per_worker_process = "4"
```

Comment:
Have you had a chance to check to see the difference when this is set to 1? It seems as though we should not really expect significant performance gains from values greater than 1 since they would all be locked on the same stream.  What really seems to matter is ensuring that we are creating a process per GPU as available. I believe this is implicit within the worker implementation, but do you know if this is true?

Reply:
Good point, I don't have the numbers for comparison of 1 vs more. I can run an existing job to see.
And yes, we're creating a worker process per GPU, that's already handled by dask-cuda.

Reply:
I saw about a 30 sec improvement with 4 over 1. The total runtime was 16mins. Removed this arg to use the default 1 thread.